modulus/models/gnn_layers/distributed_graph.py

```diff
@@ -223,6 +254,341 @@ def partition_graph_nodewise(
+    src_coordinates: torch.Tensor,
+    dst_coordinates: torch.Tensor,
+    coordinate_separators_min: List[List[float]],
+    coordinate_separators_max: List[List[float]],
+    partition_size: int,
+    partition_rank: int,
+    device: torch.device,
+) -> GraphPartition:  # pragma: no cover
+    """
+    Utility function which partitions a global graph given as CSC structure.
+    It partitions both the global ID spaces for source nodes and destination nodes
+    based on their corresponding coordinates. Each partition will manage points which
+    fulfill the boxconstraints specified by the specified coordinate separators. For each
+    rank one is expected to specify the minimum and maximum coordinate value for each dimension.
+    A partition the will manage all points for which ``min_val <= coord[d] < max_val`` holds.
+    Specifying either of these separation values as `None` will result in this constraint not being
+    considered for the division. Each rank maintains both a partition of the global source and
+    destination nodes resulting from this subspace division.
+    In terms of graph structure, each rank manages its own local graph structure
+    based on its partition of destination node IDs and all edges which - from the
+    point of view of each destination node on a current rank - are incoming edges.
+    For GNN operations this means, that features from source nodes need to be exchanged
+    between ranks. The partitioning scheme computes necessary indices which facilitate
+    later communication primitives.
+    The function performs the partitioning based on a global graph in CPU
+    memory for each rank independently. It could be rewritten to e.g. only
+    do it one rank and exchange the partitions or to an algorithm that also
+    assumes an already distributed global graph, however, we expect global
+    graphs to fit in CPU memory. After the partitioning, we can get rid off
+    the larger one in CPU memory, only keep the local graphs on each GPU, and
+    avoid tedious gather/scatter routines for exchanging partitions in the process.
+    Note: It is up to the user to ensure that the provided partition is valid.
+    In particular, we expect each rank to receive a non-empty partition of node IDs.
+
+    Examples
+    --------
+    >>> import torch
+    >>> from modulus.models.gnn_layers import partition_graph_coordinatewise
+    >>>
+    >>> # simple graph with a degree of 2 per node
+    >>> num_src_nodes = 8
+    >>> num_dst_nodes = 4
+    >>> offsets = torch.arange(num_dst_nodes + 1, dtype=torch.int64) * 2
+    >>> indices = torch.arange(num_src_nodes, dtype=torch.int64)
+    >>>
+    >>> # example with 2D coordinates
+    >>> # assuming partitioning a 2D problem into the 4 quadrants
+    >>> partition_size = 4
+    >>> partition_rank = 0
+    >>> coordinate_separators_min = [[0, 0], [None, 0], [None, None], [0, None]]
+    >>> coordinate_separators_max = [[None, None], [0, None], [0, 0], [None, 0]]
```

Comment:
How is passing `None` allowed here when the type annotation is `List[List[float]]`?